networkx/drawing/layout.py

```diff
@@ -1033,6 +1034,100 @@ def spiral_layout(G, scale=1, center=None, dim=2,
     return pos
 
 
+def multipartite_layout(G, subset_key='subset', align='vertical',
+                        scale=1, center=None):
+    """Position nodes in layers of straight lines.
+
+    Parameters
+    ----------
+    G : NetworkX graph or list of nodes
+        A position will be assigned to every node in G.
+
+    subset_key : string (default='subset')
+        Key of node data to be used as layer subset.
+
+    align : string (default='vertical')
+        The alignment of nodes. Vertical or horizontal.
+
+    scale : number (default: 1)
+        Scale factor for positions.
+
+    center : array-like or None
+        Coordinate pair around which to center the layout.
+
+    Returns
+    -------
+    pos : dict
+        A dictionary of positions keyed by node.
+
+    Examples
+    --------
+    >>> G = nx.complete_multipartite_graph(28, 16, 10)
+    >>> pos = nx.multipartite_layout(G)
+
+    Notes
+    -----
+    This algorithm currently only works in two dimensions and does not
+    try to minimize edge crossings.
+
+    Network does not need to be a complete multipartite graph. As long as nodes
+    have subset_key data, they will be placed in the corresponding layers.
+
+    """
+    import numpy as np
+
+    G, center = _process_params(G, center=center, dim=2)
+    if len(G) == 0:
+        return {}
+
+    layers = {}
+    for v, data in G.nodes(data=True):
+        try:
+            layer = data[subset_key]
+        except KeyError:
+            msg = "all nodes must have subset_key (default='subset') as data"
+            raise ValueError(msg)
+        layers[layer] = [v] + layers.get(layer, [])
+
+    pos = None
+    nodes = []
+    if align == 'vertical':
+        width = len(layers)
+        for i, layer in layers.items():
+            height = len(layer)
+            xs = np.repeat(i, height)
+            ys = np.arange(0, height, dtype=float)
+            offset = ((width-1)/2, (height-1)/2)
+            layer_pos = np.column_stack([xs, ys]) - offset
+            if pos is None:
+                pos = layer_pos
+            else:
+                pos = np.concatenate([pos, layer_pos])
+            nodes.extend(layer)
+        pos = rescale_layout(pos, scale=scale) + center
+        pos = dict(zip(nodes, pos))
+        return pos
+
+    if align == 'horizontal':
+        height = len(layers)
+        for i, layer in layers.items():
+            width = len(layer)
+            xs = np.arange(0, width, dtype=float)
+            ys = np.repeat(i, width)
+            offset =  ((width-1)/2, (height-1)/2)
+            layer_pos = np.column_stack([xs, ys]) - offset
+            if pos is None:
+                pos = layer_pos
+            else:
+                pos = np.concatenate([pos, layer_pos])
+            nodes.extend(layer)
+        pos = rescale_layout(pos, scale=scale) + center
+        pos = dict(zip(nodes, pos))
+        return pos
```

Comment:
I think this could be significantly compressed, as the if/else statements seem (if I'm reading this correctly) simply to end up only swapping the "x" and "y" positions in the end. Do you want to maybe take a stab at condensing this down to a single if/else statement (perhaps using variable names like `ax0` and `ax1` instead of `width` and `height` to help keep it clear in your head)? That would make this much easier to read (and maintain!)

I'm happy to expand on what I mean if I haven't made it clear!

Reply:
I based this implementation on the `bipartite_layout` which uses a similar approach, and which I saw as being easily readable, but I agree that there's a lot of code duplication.

Reply:
Thanks for the info, that looks worthy of a maintenance ticket :) In that case let's not make it a blocker: feel free to refactor if you want, but if not we can just include this on the maintenance ticket for another developer to take up. LMK what you prefer.